File: Algorithm/C++/971. Flip Binary Tree To Match Preorder Traversal.cc
/*
 * Created on Fri Jan 11 2019 19:46:24
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// concise dfs
// referer: https://leetcode.com/problems/flip-binary-tree-to-match-preorder-traversal/discuss/214216/JavaC++Python-DFS-Solution

#include <vector>

using std::vector;

 struct TreeNode {
   int val;
   TreeNode *left;
   TreeNode *right;
   TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
 public:
  vector<int> flipMatchVoyage(TreeNode* root, vector<int>& voyage) {
    result.clear();
    idx = 0;
    return dfs(root, voyage)? result: vector<int>{-1};
  }

  bool dfs(TreeNode* root, vector<int>& voyage) {
    if (!root) return true;
    if (root->val != voyage[idx++]) return false;
    if (root->left != NULL && root->left->val != voyage[idx]){
      result.push_back(root->val);
      return dfs(root->right, voyage) && dfs(root->left, voyage);          
    } else {
      return dfs(root->left, voyage) && dfs(root->right, voyage); 
    }
  }

 private:
  vector<int> result;
  int idx;
};

File: Algorithm/Go/971. Flip Binary Tree To Match Preorder Traversal.go
/*
 * Created on Fri Jan 11 2019 20:49:28
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// concise dfs
// referer: https://leetcode.com/problems/flip-binary-tree-to-match-preorder-traversal/discuss/214216/JavaC++Python-DFS-Solution
type State struct {
	Result []int
	Idx    int
}

func flipMatchVoyage(root *TreeNode, voyage []int) []int {
	s := State{[]int{}, 0}
	if dfs(root, voyage, &s) {
		return s.Result
	} else {
		return []int{-1}
	}
}

func dfs(root *TreeNode, voyage []int, s *State) bool {
	if root == nil {
		return true
	}
	if root.Val != voyage[s.Idx] {
		return false
	}
	s.Idx += 1
	if root.Left != nil && root.Left.Val != voyage[s.Idx] {
		s.Result = append(s.Result, root.Val)
		return dfs(root.Right, voyage, s) && dfs(root.Left, voyage, s)
	} else {
		return dfs(root.Left, voyage, s) && dfs(root.Right, voyage, s)
	}
}